test/testlib/OpenSSL/Test.pm

```diff
@@ -845,10 +843,22 @@ sub __cwd {
     foreach (@direnv) {
 	if (!file_name_is_absolute($ENV{$_})) {
 	    my $newpath = abs2rel(rel2abs($ENV{$_}), rel2abs($dir));
-	    $ENV{$_} = $newpath;
+	    $tmp_ENV{$_} = $newpath;
 	}
     }
 
+    # Should we just bail out here as well?  I'm unsure.
+    return undef unless chdir($dir);
+
+    if ($opts{cleanup}) {
+	rmtree(".", { safe => 0, keep_root => 1 });
+    }
+
+    %directories = %tmp_directories;
+    foreach (keys %tmp_ENV) {
+        $ENV{$_} = $tmp_ENV{$_};
+    }
+
```

Comment:
Say we just want to keep trick of where the source top is, so we'd have a `$directories{srctop} = ".."`, for example.

Now, consider we move into the subdirectory `foo`, and to do good, we want to make sure we know where the source top is relative to the new location.

The problem actually starts with `rel2abs`, which calculates the absolute directory for a given input.  Which result do you think would be true, doing `rel2abs($directories{srctop})` before or after doing `chdir($dir)`?